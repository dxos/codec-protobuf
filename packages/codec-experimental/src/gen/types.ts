import { MyKey } from "../my-key";
export interface Key {
    data: string;
}
export interface Task {
    id?: string;
    title?: string;
    key: MyKey;
}
export interface TaskList {
    tasks?: Task[];
}
