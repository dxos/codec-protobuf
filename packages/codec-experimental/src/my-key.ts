export class MyKey {
  constructor(
    public keyData: string,
  ) {}
}
