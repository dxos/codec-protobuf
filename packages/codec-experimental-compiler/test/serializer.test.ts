import { serializer } from './gen/serializer';
import { MyKey } from './my-key';
import { readFileSync } from 'fs'

test('encode and decode', async () => {
  const codec = serializer.getCodecForType('TaskList')

  const initial = {
    tasks: [
      {
        id: 'foo',
        title: 'Bar',
        key: new MyKey(Buffer.from('foo')),
      },
      {
        id: 'baz',
        title: 'Baz',
        key: new MyKey(Buffer.from('foo')),
      }
    ],
  }

  const encoded = codec.encode(initial)
  
  expect(encoded).toBeInstanceOf(Uint8Array);

  const decoded = codec.decode(encoded)

  expect(decoded).toEqual(initial)
})

test('definitions', () => {
  expect(readFileSync(require.resolve('./gen/serializer'), { encoding: 'utf-8' })).toMatchSnapshot()
})
