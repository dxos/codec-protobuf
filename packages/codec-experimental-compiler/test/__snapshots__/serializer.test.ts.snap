// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`definitions 1`] = `
"import { Serializer } from \\"@dxos/codec-experimental-runtime\\";
import substitutions from \\"../substitutions\\";
import { MyKey } from \\"../my-key\\";
export interface Key {
    data: string;
}
export interface AllTypes {
    doubleField?: number;
    floatField?: number;
    int32Field?: number;
    int64Field?: number;
    uint32Field?: number;
    uint64Field?: number;
    sint32Field?: number;
    sint64Field?: number;
    fixed32Field?: number;
    fixed64Field?: number;
    sfixed32Field?: number;
    sfixed64Field?: number;
    boolField?: boolean;
    stringField?: string;
    bytesField?: Uint8Array;
}
export interface Task {
    id?: string;
    title?: string;
    key: MyKey;
}
export interface TaskList {
    tasks?: Task[];
}
export const schemaJson = JSON.parse(\\"{\\\\\\"nested\\\\\\":{\\\\\\"Key\\\\\\":{\\\\\\"fields\\\\\\":{\\\\\\"data\\\\\\":{\\\\\\"rule\\\\\\":\\\\\\"required\\\\\\",\\\\\\"type\\\\\\":\\\\\\"string\\\\\\",\\\\\\"id\\\\\\":1}}},\\\\\\"AllTypes\\\\\\":{\\\\\\"fields\\\\\\":{\\\\\\"doubleField\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"double\\\\\\",\\\\\\"id\\\\\\":1},\\\\\\"floatField\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"float\\\\\\",\\\\\\"id\\\\\\":2},\\\\\\"int32Field\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"int32\\\\\\",\\\\\\"id\\\\\\":3},\\\\\\"int64Field\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"int64\\\\\\",\\\\\\"id\\\\\\":4},\\\\\\"uint32Field\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"uint32\\\\\\",\\\\\\"id\\\\\\":5},\\\\\\"uint64Field\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"uint64\\\\\\",\\\\\\"id\\\\\\":6},\\\\\\"sint32Field\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"sint32\\\\\\",\\\\\\"id\\\\\\":7},\\\\\\"sint64Field\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"sint64\\\\\\",\\\\\\"id\\\\\\":8},\\\\\\"fixed32Field\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"fixed32\\\\\\",\\\\\\"id\\\\\\":9},\\\\\\"fixed64Field\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"fixed64\\\\\\",\\\\\\"id\\\\\\":10},\\\\\\"sfixed32Field\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"sfixed32\\\\\\",\\\\\\"id\\\\\\":11},\\\\\\"sfixed64Field\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"sfixed64\\\\\\",\\\\\\"id\\\\\\":12},\\\\\\"boolField\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"bool\\\\\\",\\\\\\"id\\\\\\":13},\\\\\\"stringField\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"string\\\\\\",\\\\\\"id\\\\\\":14},\\\\\\"bytesField\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"bytes\\\\\\",\\\\\\"id\\\\\\":15}}},\\\\\\"Task\\\\\\":{\\\\\\"fields\\\\\\":{\\\\\\"id\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"string\\\\\\",\\\\\\"id\\\\\\":1},\\\\\\"title\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"string\\\\\\",\\\\\\"id\\\\\\":2},\\\\\\"key\\\\\\":{\\\\\\"rule\\\\\\":\\\\\\"required\\\\\\",\\\\\\"type\\\\\\":\\\\\\"Key\\\\\\",\\\\\\"id\\\\\\":3}}},\\\\\\"TaskList\\\\\\":{\\\\\\"fields\\\\\\":{\\\\\\"tasks\\\\\\":{\\\\\\"rule\\\\\\":\\\\\\"repeated\\\\\\",\\\\\\"type\\\\\\":\\\\\\"Task\\\\\\",\\\\\\"id\\\\\\":1}}}}}\\");
export const serializer = Serializer.fromJsonSchema(schemaJson, substitutions);
"
`;
